src/time_bucket.c

```diff
@@ -291,3 +291,63 @@ ts_time_bucket_by_type(int64 interval, int64 timestamp, Oid timestamp_type)
 
 	return ts_time_value_to_internal(time_bucketed, timestamp_type);
 }
+
+/*
+ * Add one bucket to a time value.
+ *
+ * The time value is typically the start of a time bucket, but need not be.
+ */
+Datum
+ts_time_bucket_add_one(Datum timeval, int64 bucket_width, Oid timetype)
+{
+	Datum new_timeval = timeval;
+
+	switch (timetype)
+	{
+		case DATEOID:
+		{
+			DateADT date = DatumGetDateADT(timeval);
+
+			if (date < (TS_DATE_END - bucket_width))
+				new_timeval = DateADTGetDatum(date + bucket_width);
+			break;
+		}
+		case TIMESTAMPTZOID:
+		case TIMESTAMPOID:
+		{
+			TimestampTz timestamp = DatumGetTimestampTz(timeval);
+
+			if (timestamp < (TS_TIMESTAMP_END - bucket_width))
+				new_timeval = TimestampTzGetDatum(timestamp + bucket_width);
+			break;
```

Comment:
This is slightly odd semantics, not sure if it is correct.

For times that are not in the last bucket, it adds a bucket width, but for times in the last bucket, it leaves it unchanged. Shouldn't it be a saturated add and `new_timeval` set to `TS_TIMESTAMP_END`?

Reply:
Yes, I think you are right. But I am not sure we can set it to TS_TIMESTAMP_END, because it is not a valid timestamp (end is exclusive). So probably TS_TIMESTAMP_END-1.

Reply:
Seems reasonable.